### ISSUE TITLE
Should there be a way for an `owned` class to opt into being copyable?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#16167]
This might fall into the "crazy idea" category, but it came up while looking through some code today, so I thought I'd capture it before moving on.

Non-nilable owned classes can't be "copied" through simple assignment because that results in an ownership transfer rather than a copy and there's no obvious value to store into the RHS.  Should there be a way for an author of an owned class to opt into making their class (deep) copyable?

For context, I have a list of `owned C` and would like to make a copy of it.  But if I try to do this through assignment, I get "error: Cannot copy list with element type that cannot be copied" which is reasonable given the current behavior and definition of Chapel.

I can write the deep copy of the list myself by defining and manually calling a `copy()` method on my class, but this feels fairly code-intensive.  And it means that I need to special case lists-of-owned in code that might otherwise be generic for lists-of-X.

This makes me wonder whether the presence of a standard method like `copy()` or `clone()` on a class might be a means of opting into deep-copying an owned class in cases like this rather than requiring the user to fall back to a more verbose and type-specific workaround?

Of course, class authors only define classes, not "owned classes", so there's some question about what the presence of such a method would mean for a list of unmanaged, shared, or borrowed classes...

As I said at the outset, this thought may be crazy, but I wanted to put it out there.

**Comment by mppf** [comment]
This feels to me like a different memory management discipline that is a sibling to `owned` and `shared`.

**Comment by mppf** [comment]
(I think getting the swap operator to work for this kind of case would be a reasonable alternative to the deep copies, as well).

**Comment by bradcray** [comment]
I'm not clear how the swap operator would come into play here or help.  Can you explain?

**Comment by mppf** [comment]
Say I wanted to fill in a list of non-nilable owned from another list:

``` chapel
var lstA: list(owned C) = computeList()
var lstB: list(owned C);
// want to transfer elements from `lstA` to `lstB`.
```

Here I am supposing that it is OK to destroy `lstA` in the process. If that's not true - the `copy` or `clone` strategy is more reasonable - but maybe the code should use `shared` anyway.

I could write something like this

``` chapel
// move elements from lstA to lstB, consuming them
while ! lstA.isEmpty() {
  var elt = lstA.pop();
  lstB.append(elt);
}
```

That's reasonably efficient but ends up swapping the order of the elements. That could be fixed though with another temporary list. But anyway it's pretty long.

One could also write

``` chapel
lstA <=> lstB; // swap lstA and lstB elements
```

to do the same thing; since lstB was empty, now lstA will be empty. All the elements that were in lstA will be moved to lstB. This is shorter and can be made to work with `owned C` elements by just swapping pointers to buffers and/or moving individual elements.

**Comment by bradcray** [comment]
> Here I am supposing that it is OK to destroy lstA in the process.

Oh sorry, no.  I was interested in the case of truly copying the list, not transferring it, so that the elements in either copy of the list could be messed with independently (so `shared` wouldn't work either).

**Comment by mppf** [comment]
I'm not seeing why `shared` can't work?

``` chapel
var lstA: list(shared C) = computeList()
var lstB: list(shared C);
// want to transfer elements from `lstA` to `lstB`
for elt in lstA {
  lstB.append(elt);
}
```

**Comment by mppf** [comment]
> so that the elements in either copy of the list could be messed with independently

Oh, were you talking about wanting things like `lstElement.someField = 1` only modifying the element in one list?

**Comment by bradcray** [comment]
Right.  I literally want a copy of the list, similar to how `myIntList2 = myIntList1` would give me one.

**Comment by mppf** [comment]
It seems to me then it would make more sense to have the list be of records, but if we wanted to we could create a record similar to `owned` that does deep copies. I don't think we should make `owned` do this as it seems like a departure from the unique ownership model that would deserve a different name.

**Comment by bradcray** [comment]
I can see that, yeah.  Taking your first and most recent comments together, something like having a `deepcopy C` class flavor rather than requiring each user to create their own record wrapper seems attractive.  That said, maybe it doesn't need or deserve language support as much as our other class flavors do, so maybe it should be a library-provided generic record type instead?

**Comment by mppf** [comment]
There is also the feature request to make user-defineable management types, which we could explore with this. A big part of it (other than some conventions) is marking the record implementing a management type with a pragma. But we'd have some parser work to do in order to get `new mymanagement C(20)` or `var x: mymanagement C;` to parse and I don't know how hard that would be. (Basically, we'd be generalizing parser rules expecting TOWNED etc to accept TIDENT).

**Comment by bradcray** [comment]
We could, though I was thinking that if it were just a record that didn't have or need any special compiler support, I'd be fine writing it as `list(deepcopy(C))` rather than needing special `list(deepcopy C)` syntax.
